app/src/main/java/org/mozilla/fenix/components/metrics/MozillaProductDetector.kt

```diff
@@ -27,8 +27,29 @@ object MozillaProductDetector {
         LOCKWISE("mozilla.lockbox")
     }
 
+    // The results of getInstalledMozillaProducts or getMozillaBrowserDefault can be cached. When
+    // either of those functions is called with the same context as the previous invocation, return
+    // the cached results from cachedMozillaProducts or cachedDefaultBrowserPackageName, respectively.
+    private var cachedContext: Context? = null
+    private var cachedMozillaProducts: MutableList<String>? = null
+    private var cachedDefaultBrowserPackageName: String? = null
+    private var cachedDefaultBrowserPackageNameSet: Boolean = false
```

Comment:
How does this improve startup time? afaik, at most call sites we create multiple instances but instance variable caching would not improve this behavior.

Also, I think the callers can do instance caching more simply:
```kotlin
val installedMozillaProducts = Browsers.all().getInstalledMozillaProducts(context)
```

---

fwiw, before we invest too much time into caching, which is generally complex, I'd rather ensure:
- Browsers is defined efficiently: it seems to initialize more functionality than it needs to return the results for a single method call
- We only call Browsers when necessary & we re-use the calculations from Browsers if it's used multiple times in the same place
- When possible, we defer Browsers calls to other threads

I'm not sure if we've done this already.

Reply:
Oh, this is in `MozillaProductDetector`: let me look again.

Reply:
To take a step back and look at the root cause, on startup we initialize Browsers, a slow operation, 6 times. 5 of these calls are wrapped in the MozillaProductDetector methods. They are in:
- LeanplumMetricsService 2x
- GleanMetricsService 3x
- HomeFragment.onStart calls `Settings.showPrivateModeContextualFeatureRecommender`

Furthermore, `MozillaProductDetector.getInstalledMozillaProducts` calls `packageIsInstalled`, which is slow but takes only ~15% of the time as `Browsers.all`, 3 times during startup.

To be thorough, outside of startup, `DefaultBrowserSettingsFragment` calls Browsers 3 times when shown & `IntentReceiverActivity` calls it once per processed Intent (I think that's most intents besides the homescreen intent).

---

I think the root cause is that the existing code doesn't treat `Browsers` (and its wrappers) as an expensive operation. This root cause has three PRs opened against it: this, https://github.com/mozilla-mobile/fenix/pull/7475, and https://github.com/mozilla-mobile/fenix/pull/7470. I think we should try to unify our solution.

Your patch partially addresses the root cause by adding caching in the wrapper but it can do better by also addressing the re-use of Browsers. Another issue that I see with the current implementation is that it caches the content indefinitely which may lead to incorrectness: if the user has backgrounded the app and returns, they may have uninstalled valid Mozilla products or changed their default browser. Even if the app remains open, app installations can change in the background, e.g. if the Play Store just finished installing an app or via adb.

Furthermore, after we add caching, we should question whether it's still valuable to initialize Leanplum & Glean metrics on a worker thread.

To try to sum up my thinking, **I think we should get these values once during startup and keep things on the main thread to keep it simple.** If LeanplumMetrics and GleanMetrics are still slow, we can file a follow-up issue to address it because we'll still have big wins by reducing `Browsers` calls from 6 to 1.